query/src/main/scala/filodb/query/exec/rangefn/RangeFunction.scala

```diff
@@ -68,32 +71,105 @@ trait RangeFunction {
             queryConfig: QueryConfig): Unit
 }
 
+/**
+ * Improved RangeFunction API with direct chunk access for faster columnar/bulk operations
+ */
+trait ChunkedRangeFunction extends RangeFunction {
+  def addedToWindow(row: TransientRow, window: Window): Unit = {}
+
+  def removedFromWindow(row: TransientRow, window: Window): Unit = {}
+
+  def apply(startTimestamp: Long, endTimestamp: Long, window: Window,
+            sampleToEmit: TransientRow,
+            queryConfig: QueryConfig): Unit = {}
+
+  /**
+   * Resets the state
+   */
+  def reset(): Unit = {}
+
+  /**
+   * Called by the ChunkedWindowIterator to add multiple rows at a time to the range function for efficiency.
+   * The idea is to call chunk-based methods such as sum and binarySearch.
+   * @param tsCol ColumnID for timestamp column
+   * @param valueCol ColumnID for value column
+   * @param info ChunkSetInfo with information for specific chunks
+   * @param startTime starting timestamp in millis since Epoch for time window
+   */
+  def addChunks(tsCol: Int, valueCol: Int, info: ChunkSetInfo,
+                startTime: Long, endTime: Long, queryConfig: QueryConfig): Unit
+
+  /**
+   * Return the computed result in the sampleToEmit
+   */
+  def apply(endTimestamp: Long, sampleToEmit: TransientRow): Unit
+}
+
+/**
+ * Standard ChunkedRangeFunction implementation extracting the start and ending row numbers from the timestamp
+ * and returning the double value vector and reader with the row numbers
+ */
+trait ChunkedDoubleRangeFunction extends ChunkedRangeFunction {
+  def addChunks(tsCol: Int, valueCol: Int, info: ChunkSetInfo,
+                startTime: Long, endTime: Long, queryConfig: QueryConfig): Unit = {
+    val timestampVector = info.vectorPtr(tsCol)
+    val tsReader = bv.LongBinaryVector(timestampVector)
+    val doubleVector = info.vectorPtr(valueCol)
+    val dblReader = bv.DoubleVector(doubleVector)
+
+    // First row >= startTime, so we can just drop bit 31 (dont care if it matches exactly)
+    val startRowNum = tsReader.binarySearch(timestampVector, startTime) & 0x7fffffff
+    val endRowNum = tsReader.ceilingIndex(timestampVector, endTime)
```

Comment:
I see that the above few lines will be a code pattern used in multiple places - for example, in downsampling. Would be nice to abstract this kind of a search further. Please see if possible. 

This is where cursors would really be beneficial.

Reply:
I agree cursors are the right way to go.  I think it's better to leave the abstraction and effort to cursors as that will abstract all those things.